art/experiment/Experiment.py

```diff
@@ -24,45 +24,68 @@ def __init__(self, name, **kwargs):
         self.steps.append(step)
-        self.state.step_states[step.get_model_name()][step.get_name_with_id()] = step._get_saved_state()
         step.set_step_id(len(self.steps))
-        step.set_experiment(self)
         self.checks.append(checks)
 
+    def fill_step_states(self, step):
+        self.state.step_states[step.get_model_name()][
+            step.get_name_with_id()
+        ] = step.get_results()
+
     def run_all(self):
         MetricCalculator.set_experiment(self)
         MetricCalculator.create_exceptions()
+
         for step, checks in zip(self.steps, self.checks):
             self.state.current_step = step
-            step_passed = True
 
-            for check in checks:
-                result = check.check(step)
-                if not result.is_positive:
-                    step_passed = False
-                    break
-
-            model_changed = False
+            step_passed = True
+            if not step.was_run():  # fmt: skip
+                step_passed = False
+            else:
+                step.load_results()
+                for check in checks:
+                    check_result = check.check(step)
+                    if not check_result.is_positive:
+                        step_passed = False
+                        break
 
             if step_passed:
                 step_current_hash = step.get_hash()
                 step_saved_hash = step.get_results()["hash"]
+                model_changed = False
                 if step_current_hash != step_saved_hash:
                     model_changed = True
                 if not model_changed:
-                    print(f"Step {step.name}_{step.get_step_id()} was already completed.")
+                    print(
+                        f"Step {step.name}_{step.get_step_id()} was already completed."
+                    )
+                    self.fill_step_states(step)
                     continue
                 else:
-                    print(f"We've noticed you changed some code in your model in step {step.name}_{step.get_step_id()}. We will rerun it.")
+                    print(
+                        f"We've noticed you changed some code in your model in step {step.name}_{step.get_step_id()}. We will rerun it."
+                    )
 
             step.add_result("hash", step.get_hash())
 
             step(self.state.step_states)
             for check in checks:
                 result = check.check(step)
                 if not result.is_positive:
-                    raise Exception(f"Check failed for step: {step.name}. Reason: {result.error}")
+                    raise Exception(
+                        f"Check failed for step: {step.name}. Reason: {result.error}"
+                    )
+
+            self.fill_step_states(step)
+
         self.logger = None
 
     def get_steps(self):
         return self.steps
+
+    def exchange_step(self, step: "Step", step_id=-1):
```

Comment:
what does it do? I don't know if `exchange_step` is a good name for it, as I see it in https://github.com/SebChw/art_template/pull/4/files

Reply:
Just look on how it is used within tutorial. I don't know if it's the best idea, but for sure being able to exchange some step with other one seemed to me like a useful feature

Reply:
I think that this may be useful If someone would try to perform step in a loop until it is passed for different parameters. Then this may be handy